src/likelihoods/negativebinomial.jl

```diff
@@ -1,36 +1,169 @@
-    Note that with symmetric functions like the [`LogisticLink`](@ref), this corresponds to
-    using `l(-f)`.
-"""
-struct NegativeBinomialLikelihood{Tl<:AbstractLink,T<:Real} <: AbstractLikelihood
-    successes::T    # number of successes parameter
+
+[^1] Hilbe, Joseph M. Negative binomial regression. Cambridge University Press, 2011.
+"""
+struct NegativeBinomialLikelihood{Tp<:NBParam,Tl<:AbstractLink} <: AbstractLikelihood
+    params::Tp # Likelihood parametrization (and parameters)
     invlink::Tl
+    function NegativeBinomialLikelihood(
+        params::Tparam, invlink=default_invlink(params)
+    ) where {Tparam<:NBParam}
+        # we convert `invlink` into `Link` if it's not the case already
+        invlink = link(invlink)
+        return new{Tparam,typeof(invlink)}(params, invlink)
+    end
 end
 
-function NegativeBinomialLikelihood(l=logistic; successes::Real=1)
-    return NegativeBinomialLikelihood(successes, link(l))
+@deprecate NegativeBinomialLikelihood(l=logistic; successes=1) NegativeBinomialLikelihood(
+    NBParamI(successes), l
+)
+
+function (l::NegativeBinomialLikelihood)(::Real)
+    return error(
+        "not implemented for type $(typeof(l)). For your custom type to run you ",
+        "need to implement `(l::NegativeBinomialLikelihood{<:MyNBParam})(f::Real)`. ",
+        "For a full explanation, see `NegativeBinomialLikelihood` docs",
+    )
 end
 
 @functor NegativeBinomialLikelihood
 
-(l::NegativeBinomialLikelihood)(f::Real) = NegativeBinomial(l.successes, l.invlink(f))
-
 (l::NegativeBinomialLikelihood)(fs::AbstractVector{<:Real}) = Product(map(l, fs))
+
+"""
+    NBParamSuccess(successes)
+
+Negative Binomial parametrization with `successes` the number of successes and
+`link(f)` the probability of `success`.
+This corresponds to the definition used by [Distributions.jl](https://juliastats.org/Distributions.jl/latest/univariate/#Distributions.NegativeBinomial).
+
+```math
+  p(k|successes, f) = \\frac{\\Gamma(k+successes)}{k! \\Gamma(successes)} l(f)^successes (1 - l(f))^k
+```
+"""
+struct NBParamSuccess{T} <: NBParamProb
+    successes::T
+end
+
+function (l::NegativeBinomialLikelihood{<:NBParamSuccess})(f::Real)
+    return NegativeBinomial(l.params.successes, l.invlink(f))
+end
+
+"""
+    NBParamFailure(failures)
+
+Negative Binomial parametrization with `failures` the number of failures and
+`l(f)` the probability of `success`.
+This corresponds to the definition used by [Wikipedia](https://en.wikipedia.org/wiki/Negative_binomial_distribution).
+
+```math
+  p(k|failures, f) = \\frac{\\Gamma(k+failures)}{k! \\Gamma(failure)} l(f)^k (1 - l(f))^{failures}
+```
+"""
+struct NBParamFailure{T} <: NBParamProb
+    failures::T
+end
+
+function (l::NegativeBinomialLikelihood{<:NBParamFailure})(f::Real)
+    return NegativeBinomial(l.params.failures, 1 - l.invlink(f))
+end
+
+# Helper function to convert mean and variance to p and r
+_nb_mean_var_to_r_p(μ::Real, v::Real) = abs2(μ) / (v - μ), μ / v
```

Comment:
```suggestion
_nb_mean_excessvar_to_r_p(μ::Real, ev::Real) = abs2(μ) / ev, μ / (μ + ev)
```
to be fair I've not tried out how much of a difference it makes, but I'm always a bit concerned about `a + b - a` if we can avoid it...

Reply:
Where do you see a `(a + b - a)` ?

Reply:
Ah but what you propose makes sense.

Reply:
v - μ, where v = μ + μ * ev, so I changed it to have all the cancellations disappear analytically instead of numerically